modules/core/src/smithy4s/http/internals/HttpResponseCodeSchemaVisitor.scala

```diff
@@ -0,0 +1,141 @@
+  def default[A]: ResponseCodeExtractor[A] = NoResponseCode
+  private val primPf = Primitive.deriving[ResponseCodeExtractor]
+  override def primitive[P](
+      shapeId: ShapeId,
+      hints: Hints,
+      tag: Primitive[P]
+  ): ResponseCodeExtractor[P] = tag match {
+    case PInt =>
+      if (hints.has[smithy.api.HttpResponseCode]) {
+        primPf(tag)
+      } else {
+        NoResponseCode
+      }
+    case _ => NoResponseCode
+  }
+
+  override def struct[S](
+      shapeId: ShapeId,
+      hints: Hints,
+      fields: Vector[SchemaField[S, _]],
+      make: IndexedSeq[Any] => S
+  ): ResponseCodeExtractor[S] = {
+    def compileField[A](
+        field: SchemaField[S, A]
+    ): Option[ResponseCodeExtractor[S]] = {
+      val folder = new Field.Folder[Schema, S, ResponseCodeExtractor[S]]() {
+        def onOptional[AA](
+            label: String,
+            instance: Schema[AA],
+            get: S => Option[AA]
+        ): ResponseCodeExtractor[S] = {
+          val aExt = apply(instance)
+          OptionalResponseCode { s =>
+            get(s) match {
+              case None => None
+              case Some(value) =>
+                aExt match {
+                  case NoResponseCode          => None
+                  case RequiredResponseCode(f) => Some(f(value))
+                  case OptionalResponseCode(f) => f(value)
+                }
+            }
+          }
+        }
+        def onRequired[AA](
+            label: String,
+            instance: Schema[AA],
+            get: S => AA
+        ): ResponseCodeExtractor[S] = {
+          val aExt = apply(instance)
+          Contravariant[ResponseCodeExtractor].contramap(aExt)(get)
+        }
+      }
+
+      field.hints
+        .get[smithy.api.HttpResponseCode]
+        .map(_ => field.fold(folder))
+    }
+    fields.flatMap(f => compileField(f)).headOption.getOrElse(NoResponseCode)
+  }
+}
+
+object HttpResponseCodeSchemaVisitor {
+  sealed trait ResponseCodeExtractor[-A]
+  object NoResponseCode extends ResponseCodeExtractor[Any]
+  case class RequiredResponseCode[A](f: A => Int)
+      extends ResponseCodeExtractor[A]
+  case class OptionalResponseCode[A](f: A => Option[Int])
+      extends ResponseCodeExtractor[A]
+
+  implicit val contravariant: Contravariant[ResponseCodeExtractor] =
+    new Contravariant[ResponseCodeExtractor]() {
+      def contramap[A, B](
+          fa: ResponseCodeExtractor[A]
+      )(f: B => A): ResponseCodeExtractor[B] = {
+        fa match {
+          case NoResponseCode          => NoResponseCode
+          case RequiredResponseCode(g) => RequiredResponseCode { b => g(f(b)) }
+          case OptionalResponseCode(g) => OptionalResponseCode { b => g(f(b)) }
+        }
+      }
+    }
+
+  implicit val short: ResponseCodeExtractor[Short] = NoResponseCode
+  implicit val int: ResponseCodeExtractor[Int] = {
+    RequiredResponseCode(identity)
+  }
+  implicit val float: ResponseCodeExtractor[Float] = NoResponseCode
```

Comment:
I think you shouldn't need to define them. `Primitive.deriving` exists to reduce the boilerplate when interop-ing with existing libraries that are likely to define these instances. Plus you're already pattern-matching against `int` in the implementation of the `primitive` method.